network/ocsp.go

```diff
@@ -0,0 +1,232 @@
+	"io/ioutil"
+	"net/http"
+	"net/url"
+)
+
+const (
+	ocspRequiredYes int = iota
+	ocspRequiredNo
+	ocspRequiredAll
+)
+
+const (
+	ocspFromCertUse int = iota
+	ocspFromCertTrust
+	ocspFromCertPrefer
+	ocspFromCertIgnore
+)
+
+type OCSPConfig struct {
+	url      string
+	required int // ocspRequired*
+	fromCert int // ocspFromCert*
+}
+
+func NewOCSPConfig(uri, required, fromCert string) (*OCSPConfig, error) {
+	if len(uri) > 0 {
+		_, err := url.Parse(uri)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	var requiredVal int
+	switch required {
+	case "yes", "true":
+		requiredVal = ocspRequiredYes
+	case "no", "false":
+		requiredVal = ocspRequiredNo
+	case "all":
+		requiredVal = ocspRequiredAll
+	default:
+		return nil, errors.New("Invalid `ocsp_required` value '" + required + "', should be one of 'yes', 'no', 'all'")
+	}
+
+	var fromCertVal int
+	switch fromCert {
+	case "use":
+		fromCertVal = ocspFromCertUse
+	case "trust":
+		fromCertVal = ocspFromCertTrust
+	case "prefer":
+		fromCertVal = ocspFromCertPrefer
+	case "ignore":
+		fromCertVal = ocspFromCertIgnore
+	default:
+		return nil, errors.New("Invalid `ocsp_from_cert` value '" + fromCert + "', should be one of 'use', 'trust', 'prefer', 'ignore'")
+	}
+
+	if len(uri) > 0 {
+		log.Debugf("OCSP: Using server '%s'", uri)
+	}
+
+	switch required {
+	case "yes", "true":
+		log.Debugf("OCSP: At least one OCSP server should confirm certificate validity")
+	case "no", "false":
+		log.Debugf("OCSP: Allowing certificates not known by OCSP server")
+	case "all":
+		log.Debugf("OCSP: Requiring positive response from all OCSP servers")
+	}
+
+	switch fromCert {
+	case "use":
+		log.Debugf("OCSP: using servers described in certificates if nothing passed via command line")
+	case "trust":
+		log.Debugf("OCSP: trusting responses from OCSP servers listed in certificates")
+	case "prefer":
+		log.Debugf("OCSP: server from certificate will be prioritized over one from command line")
+	case "ignore":
+		log.Debugf("OCSP: ignoring OCSP servers described in certificates")
+	}
+
+	return &OCSPConfig{url: uri, required: requiredVal, fromCert: fromCertVal}, nil
+}
+
+func (c *OCSPConfig) Describe() string {
+	return fmt.Sprintf("url=%s, required=%d, fromCert=%d", c.url, c.required, c.fromCert)
+}
+
+func queryOCSP(commonName string, clientCert, issuerCert *x509.Certificate, ocspServer string) (*ocsp.Response, error) {
+	opts := &ocsp.RequestOptions{Hash: crypto.SHA256}
+	buffer, err := ocsp.CreateRequest(clientCert, issuerCert, opts)
+	if err != nil {
+		return nil, err
+	}
+	httpRequest, err := http.NewRequest(http.MethodPost, ocspServer, bytes.NewBuffer(buffer))
+	if err != nil {
+		return nil, err
+	}
+	ocspUrl, err := url.Parse(ocspServer)
+	if err != nil {
+		return nil, err
+	}
+	httpRequest.Header.Add("Content-Type", "application/ocsp-request")
+	httpRequest.Header.Add("Accept", "application/ocsp-response")
+	httpRequest.Header.Add("host", ocspUrl.Host)
+	httpClient := &http.Client{}
```

Comment:
omg, one more TLS config, to query OCSP over HTTPS, and it probably needs its own OCSP config

Reply:
Hm... Isn't this prone to recursion? IIRC, OCSP servers usually work over HTTP exactly for this reason: so that you don't have to special-case TLS connections to OCSP servers to skip OCSP checks for themselves.

Reply:
I missed, that response of ocsp should be signed. So we will validate it by key. Plus ocsp request contains only hashes without any information about certificates (but allow to use tables of hashes of public certificates for matching...). So actually yes, we can request it by http.

Reply:
here the same advice, to move client initialization to constructor with pre-defined timeouts/transport object.